### PR TITLE
Expose partial sync API

### LINKED ISSUE #1565: Add support for partial sync
Product: https://github.com/realm/product/issues/245
OS: https://github.com/realm/realm-object-store/pull/547

### PR DESCRIPTION
**Pull request by nirinchev**
## Description

Fixes #1565 

##  TODO

* [x] Changelog entry
* [x] Docs
* [x] Tests 
* [x] Update PCL

## TODO after merging
- [ ] Expose string-based `.Where` method on IQueryable. This will allow people to run the query locally.

### CODE REVIEW

File: Realm/Realm.Sync/Handles/SharedRealmHandleExtensions.cs [hunk 1]
```diff
@@ -57,6 +57,9 @@ private static class NativeMethods
             [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
             public unsafe delegate void SessionWaitCallback(IntPtr task_completion_source, int error_code, byte* message_buf, IntPtr message_len);
 
+            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+            public unsafe delegate void SubscribeForObjectsCallback(IntPtr results, IntPtr task_completion_source, NativeException ex);
```
Comment (fealebenpae) [inline Realm/Realm.Sync/Handles/SharedRealmHandleExtensions.cs]: This doesn't need to be `unsafe`.

File: Realm/Realm.Sync/Handles/SharedRealmHandleExtensions.cs [hunk 2]
```diff
@@ -66,6 +69,9 @@ private static class NativeMethods
             [DllImport(InteropConfig.DLL_NAME, EntryPoint = "realm_install_syncsession_callbacks", CallingConvention = CallingConvention.Cdecl)]
             public static extern unsafe void install_syncsession_callbacks(RefreshAccessTokenCallbackDelegate refresh_callback, SessionErrorCallback error_callback, SessionProgressCallback progress_callback, SessionWaitCallback wait_callback);
 
+            [DllImport(InteropConfig.DLL_NAME, EntryPoint = "realm_syncmanager_install_callbacks", CallingConvention = CallingConvention.Cdecl)]
+            public static extern unsafe void install_callbacks(SubscribeForObjectsCallback subscribe_callback);
```
Comment (fealebenpae) [inline Realm/Realm.Sync/Handles/SharedRealmHandleExtensions.cs]: Ditto.

File: Realm/Realm.Sync/Handles/SharedRealmHandleExtensions.cs [hunk 3]
```diff
@@ -273,5 +298,30 @@ private static unsafe void HandleSessionWaitCallback(IntPtr taskCompletionSource
                 handle.Free();
             }
         }
+
+        [NativeCallback(typeof(NativeMethods.SubscribeForObjectsCallback))]
+        private static unsafe void HandleSubscribeForObjectsCallback(IntPtr results, IntPtr taskCompletionSource, NativeException ex)
```
Comment (fealebenpae) [inline Realm/Realm.Sync/Handles/SharedRealmHandleExtensions.cs]: No need for `unsafe`.

File: Realm/Realm.Sync/Native/SyncConfiguration.cs [hunk 4]
```diff
@@ -50,7 +50,7 @@ internal string Url
         }
 
         [MarshalAs(UnmanagedType.I1)]
-        internal bool client_validate_ssl;
+        internal bool EnableSSLValidation;
```
Comment (fealebenpae) [inline Realm/Realm.Sync/Native/SyncConfiguration.cs]: Why the rename?

File: Realm/Realm.Sync/Native/SyncConfiguration.cs [hunk 5]
```diff
@@ -64,5 +64,21 @@ internal string TrustedCAPath
                 trusted_ca_path_len = (IntPtr)(value?.Length ?? 0);
             }
         }
+
+        [MarshalAs(UnmanagedType.I1)]
+        internal bool IsPartial;
```
Comment (fealebenpae) [inline Realm/Realm.Sync/Native/SyncConfiguration.cs]: The field names of the marshaling struct should match those in the C++ struct.

File: Tests/Tests.Sync.Shared/PartialSyncTests.cs [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+                        original.Add(new ObjectA
+                        {
+                            StringValue = "A #" + i,
+                            IntValue = i,
+                            B = new ObjectB
+                            {
+                                StringValue = "B #" + i,
+                                BoolValue = i % 2 == 0,
+							}
+						});
```
Comment (fealebenpae) [inline Tests/Tests.Sync.Shared/PartialSyncTests.cs]: Whitespace.

File: Tests/Tests.Sync.Shared/PartialSyncTests.cs [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+
+                await original.GetSession().WaitForUploadAsync();
+            }
+
+            Realm.DeleteRealm(config);
+
+            config = new SyncConfiguration(config.User, config.ServerUri, config.DatabasePath + "_partial")
+            {
+                ObjectClasses = config.ObjectClasses,
+				IsPartial = true
```
Comment (fealebenpae) [inline Tests/Tests.Sync.Shared/PartialSyncTests.cs]: Whitespace.

File: wrappers/src/sync_manager_cs.cpp [hunk 8]
```diff
@@ -193,6 +210,33 @@ REALM_EXPORT void realm_syncmanager_set_feature_token(const uint16_t* token_buf,
...
+REALM_EXPORT void realm_syncmanager_subscribe_for_objects(SharedRealm& sharedRealm, uint16_t* class_buf, size_t class_len, uint16_t* query_buf, size_t query_len, void* task_completion_source, NativeException::Marshallable& ex)
+{
+    handle_errors(ex, [&]() {
+        Utf16StringAccessor class_name(class_buf, class_len);
+        Utf16StringAccessor query(query_buf, query_len);
+
+        partial_sync::register_query(sharedRealm, class_name, query, [=](Results results, std::exception_ptr err) {
+            if (err) {
+                try {
+                    throw err;
```
Comment (fealebenpae) [inline wrappers/src/sync_manager_cs.cpp]: I believe you should use `std::rethrow_exception(err)`.

File: Realm/Realm.Sync/Native/SyncConfiguration.cs (hunk 4 above)
Comment (nirinchev) [reply Realm/Realm.Sync/Native/SyncConfiguration.cs]: Just bugged me a little that half the public members are pascal case, the others - snake case. Will revert though, as you're right - better to stick to the c++ naming for matching fields.

**Pull request merged**